pkg/controller/job/job_controller.go

```diff
@@ -861,7 +867,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 		}
 	}
 
-	if isIndexedJob(&job) {
+	if successCriteriaMetCond == nil && isIndexedJob(&job) {
```

Comment:
With the `isSuccessCriteriaMetCondition` function I think we don't need the `successCriteriaMetCond` variable (unless I'm missing something). I would prefer to have a single source of through about the "finished condition".

I imagine that we have:

```golang
...
  exceedsBackoffLimit := jobCtx.failed > *job.Spec.BackoffLimit
  finishedCondition := successCriteriaMetCondition(&job)

  if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
    // Given that the Job already has the SuccessCriteriaMet condition, the termination condition already had confirmed in another cycle.
    // So, the job-controller evaluates the podFailurePolicy only when the Job doesn't have the SuccessCriteriaMet condition.
    if finishedCondition == nil && feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
	
...

  if !isSuccessCriteriaMetCondition(jobCtx.finishedCondition) && isIndexedJob(&job) {
  ...
```

Reply:
Yes, we can replace here with `isSuccessCriteriaMetCondition(jobCtx.finishedCcondition)`, but I think that we still need the `successCriteriaMetCond ` here: https://github.com/kubernetes/kubernetes/blob/c2faf6bcacd180fcc38689b8ebe9bf574f232e7d/pkg/controller/job/job_controller.go#L1166-L1169

Because we need to avoid adding the `SuccessCriteriaMet` condition into Job twice or more.

@mimowo WDYT?

Reply:
I don't see `successCriteriaMetCond` used in the snippet you reference. I still think `successCriteriaMetCond` should not be needed as additional var, because it can be derived from `jobCtx.finishedCondition`. So, my preference is to have a single source of truth. I think you can derive it by just jobCtx.finishedCondition != nil && jobCtx.finishedCondition.Type == SuccessCriteriaMet. 

We do something similar for FailureTarget condition.

Reply:
Are you removing `successCriteriaMetCond == nil &&`? It looks like it can go away.

Reply:
> Are you removing successCriteriaMetCond == nil &&? It looks like it can go away.

@alculquicondor I will remove that conditional, and then I will introduce a new conditional, `!isSuccessCriteriaCondition`.

Reply:
Done.

Reply:
> > Are you removing successCriteriaMetCond == nil &&? It looks like it can go away.
> 
> @alculquicondor I will remove that conditional, and then I will introduce a new conditional, `!isSuccessCriteriaCondition`.

As I discussed in https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1516065458 with Michal, I removed the `successCriteriaMetCond == nil &&` from here.